system/jlib/jsecrets.cpp

```diff
@@ -0,0 +1,702 @@
+    splitUrlSections(url, authority, authorityLen, path, &schemeHostPort);
+    splitUrlAuthority(authority, authorityLen, user, password, schemeHostPort, nullptr);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+
+static StringBuffer secretDirectory;
+static CriticalSection secretCS;
+
+//there are various schemes for renewing kubernetes secrets and they are likely to vary greatly in how often
+//  a secret gets updated this timeout determines the maximum amount of time before we'll pick up a change
+//  10 minutes for now we can change this as we gather more experience and user feedback
+static unsigned secretTimeoutMs = 10 * 60 * 1000;
+
+extern jlib_decl unsigned getSecretTimeout()
+{
+    return secretTimeoutMs;
+}
+
+extern jlib_decl void setSecretTimeout(unsigned timeoutMs)
+{
+    secretTimeoutMs = timeoutMs;
+}
+
+extern jlib_decl void setSecretMount(const char * path)
+{
+    if (!path)
+    {
+        getPackageFolder(secretDirectory);
+        addPathSepChar(secretDirectory).append("secrets");
+    }
+    else
+        secretDirectory.set(path);
+}
+
+static inline bool checkSecretExpired(unsigned created)
+{
+    if (!created)
+        return false;
+    unsigned age = msTick() - created;
+    return age > getSecretTimeout();
+}
+
+class CVault
+{
+private:
+    bool useKubernetesAuth = true;
+    CVaultKind kind;
+    CriticalSection vaultCS;
+    Owned<IPropertyTree> cache;
+
+    StringBuffer schemeHostPort;
+    StringBuffer path;
+    StringBuffer username;
+    StringBuffer password;
+    StringAttr name;
+    StringAttr role;
+    StringAttr token;
+
+public:
+    CVault(IPropertyTree *vault)
+    {
+        cache.setown(createPTree());
+        StringBuffer url;
+        replaceEnvVariables(url, vault->queryProp("@url"), false);
+        if (url.length())
+            splitUrlSchemeHostPort(url.str(), username, password, schemeHostPort, path);
+        name.set(vault->queryProp("@name"));
+        kind = getSecretType(vault->queryProp("@kind"));
+        if (vault->hasProp("@role"))
+            role.set(vault->queryProp("@role"));
+        else
+            role.set("hpcc-vault-access");
+        if (vault->hasProp("@client-secret"))
+        {
+            useKubernetesAuth = false;
+            //for now only support direct access token.  we can support other combinations for example login token, ldap login, etc later.
+            Owned<IPropertyTree> clientSecret = getLocalSecret(vault->queryProp("@client-secret"));
+            if (clientSecret)
+                token.set(clientSecret->queryProp("token"));
+        }
+    }
+    CVaultKind getVaultKind() const { return kind; }
+    void kubernetesLogin()
+    {
+        CriticalBlock block(vaultCS);
+        if (token.length())
+            return;
+        StringBuffer login_token;
+        login_token.loadFile("/var/run/secrets/kubernetes.io/serviceaccount/token");
```

Comment:
Does this require the pods to have any access rights?  Some recent PRs of Richard's have tied down both the access rights and the network access.  Does that introduce problems here?

Reply:
Based on a quick comment from @richardkchapman I think the line in the README where I do:

vault write auth/kubernetes/role/hpcc-vault-access \
        bound_service_account_names=hpcc-default \
        bound_service_account_namespaces=default \
        policies=hpcc-kv-ro \
        ttl=24h
may be enough to grant access.
It's more of an operational thing on the vault side granting access to our default account, or whatever the user may change the account to.  
I also support passing in the authentication token as a k8s secret.  Hopefully that will cover some other authentication schemes the user may set up.  The user can take whatever steps necessary to generate that.
Vault supports many authentication schemes and we can add more over time based on what scenarios users need, or we want to support in general.